File: stage1/gradle-plugin-plugin/src/main/java/net/rubygrapefruit/plugins/stage1/GradlePluginPlugin.java
package net.rubygrapefruit.plugins.stage1;

import net.rubygrapefruit.plugins.stage0.BuildConstants;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.plugin.devel.GradlePluginDevelopmentExtension;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.kotlin.gradle.dsl.KotlinProjectExtension;

@SuppressWarnings("unused")
public class GradlePluginPlugin implements Plugin<Project> {
    @Override
    public void apply(@NotNull Project target) {
        target.getPlugins().apply(JvmBasePlugin.class);
        target.getPlugins().apply("java-gradle-plugin");

        target.getDependencies().add("implementation", BuildConstants.constants.stage0.buildConstants.coordinates);

        GradlePluginDevelopmentExtension pluginDevExtension = target.getExtensions().getByType(GradlePluginDevelopmentExtension.class);
        target.getExtensions().create("pluginBundle", PluginBundle.class, pluginDevExtension);
    }
}

File: local-plugins/base-plugins/src/main/kotlin/net/rubygrapefruit/plugins/app/internal/plugins/GradlePluginPlugin.kt
package net.rubygrapefruit.plugins.app.internal.plugins

import net.rubygrapefruit.plugins.app.PluginBundle
import net.rubygrapefruit.plugins.app.Versions
import net.rubygrapefruit.plugins.app.internal.DefaultPluginBundle
import net.rubygrapefruit.plugins.app.internal.componentRegistry
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.plugin.devel.GradlePluginDevelopmentExtension

@Suppress("unused")
class GradlePluginPlugin : Plugin<Project> {
    override fun apply(target: Project) {
        with(target) {
            applyBasePlugin()

            plugins.apply("java-gradle-plugin")
            plugins.apply("org.jetbrains.kotlin.jvm")
            plugins.apply(LibraryBasePlugin::class.java)
            plugins.apply(JvmConventionsPlugin::class.java)

            repositories.mavenCentral()
            repositories.gradlePluginPortal()

            dependencies.add("implementation", Versions.libs.coordinates("build-constants"))

            val pluginDevExtension = extensions.getByType(GradlePluginDevelopmentExtension::class.java)
            val bundle = extensions.create(
                PluginBundle::class.java,
                "pluginBundle",
                DefaultPluginBundle::class.java,
                pluginDevExtension
            ) as DefaultPluginBundle
            bundle.targetJvmVersion.set(Versions.plugins.jvm.version)

            componentRegistry.register(bundle)
        }
    }
}